src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrPusher.cs

```diff
@@ -0,0 +1,216 @@
+        {
+            ExecutePush(remote, branch);
+        }
+        else
+        {
+            _logger.LogError("Cannot push to the VMR");
+        }  
+    }
+
+    private void ExecutePush(
+        string remoteName,
+        string branchName)
+    {
+        using var vmrRepo = new Repository(_vmrInfo.VmrPath);
+        vmrRepo.Config.Add("user.name", Constants.DarcBotName);
+        vmrRepo.Config.Add("user.email", Constants.DarcBotEmail);
+
+        var remote = vmrRepo.Network.Remotes[remoteName];
+
+        Branch branch;
+        if (!vmrRepo.Branches.Any(b => b.FriendlyName == branchName))
+        {
+            branch = vmrRepo.CreateBranch(branchName);
+        }
+        else
+        {
+            branch = vmrRepo.Branches[branchName];
+        }
+
+        var pushOptions = new PushOptions
+        {
+            CredentialsProvider = (url, user, cred) =>
+                new UsernamePasswordCredentials
+                {
+                    Username = _vmrRemoteConfiguration.GitHubToken,
+                    Password = ""
+                }
+        };
+
+        vmrRepo.Network.Push(remote, branch.CanonicalName, pushOptions);
+    }
+
+    private async Task<bool> ValidateCommits(string gitHubApiPat, CancellationToken cancellationToken)
+    {
+        var commitsSearchArguments = _sourceManifest
+            .Repositories
+            .Select(r => GetCommitSearchArguments(r.RemoteUri, r.CommitSha));
+
+        var commits = await GetRepositoriesCommits(commitsSearchArguments, gitHubApiPat, cancellationToken);
+        
+        if (commits == null || commits.Data == null)
+        {
+            _logger.LogError("The Graphql query is not in the correct format.");
+            return false;
+        }
+
+        foreach (var pair in commits.Data)
+        {
+            if (pair.Value == null)
+            {
+                var commitArgs = commitsSearchArguments.First(r => SanitizeName(r.RepoName) == pair.Key);
+                _logger.LogError($"Repository {commitArgs.RepoName} with owner {commitArgs.RepoOwner} couldn't be found on GitHub.");
+                return false;
+            }
+
+            if (pair.Value.Object == null)
+            {
+                var commitArgs = commitsSearchArguments.First(r => SanitizeName(r.RepoName) == pair.Key);
+                _logger.LogError($"The commit {commitArgs.Sha} couldn't be found in public repo {commitArgs.RepoName} with owner {commitArgs.RepoOwner} on GitHub.");
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private async Task<CommitsQueryResult?> GetRepositoriesCommits(
+        IEnumerable<CommitSearchArguments> commits,
+        string gitHubApiPat,
+        CancellationToken cancellationToken)
+    {
+        var queries = commits
+            .Select(c => $@"{SanitizeName(c.RepoName)}: repository(name: \""{c.RepoName}\"", owner: \""{c.RepoOwner}\""){{object(oid: \""{c.Sha}\""){{ ... on Commit {{id}}}}}}");
```

Comment:
Do you need to escape `\"` when you use `@`?

Reply:
yes because it's escaping it for json - it should be \" but because it's in c# we also escape the quotes so it becomes \""

Reply:
yes, it's hard to tell with so many levels of escapes but it doesn't work without it

Reply:
I was only thinking we could remove `@` and use just one kind of escaping which might make it clearer

Reply:
I'll try it

Reply:
I would suggest using raw string literals to do this. It requires no escaping at all.

Reply:
That would be the best but we're on .NET 6.0 in this repo still